src/controllers/account/index.ts

```diff
@@ -1,45 +1,68 @@
     const { email } = ctx.request.body
-    ctx.assert(email, 400)
-    // TODO see if email address is in DB, if not: bail
-    // const resetToken = await AccountUtilities.generateResetToken()
-    // const verifierHash = await AccountUtilities.getVerifierHash(resetToken)
-    // await AccountUtilities.storeResetToken(resetToken.selector, verifierHash)
-    // const emailToken = await AccountUtilities.getEmailResetToken(resetToken)
-    // // TODO email token to user
+    ctx.assert(email, 400, Strings.EmailIsRequired)
 
-    ctx.status = 200
-    ctx.body = { ok: true }
-  }
-  public static async receiveReset (ctx: BaseContext) {
-    const { token } = ctx.request.body
-    ctx.assert(token, 400)
-    // const resetToken = await AccountUtilities.splitSelectorVerifier(token)
-    // DB lookup first 16 bytes provided (selector)
-    // if found, hash next 16 bytes provided (verifier)
-    // const verifierHash = await AccountUtilities.getVerifierHash(resetToken)
-    // compare hash with verifier hash in DB
+    // Generate token, store it
+    const emailToken = await generateAndStoreToken(ctx, false)
     try {
-      // const storedVerifierHash = await AccountUtilities.getStoredVerifierHash(resetToken)
-      // const match = await AccountUtilities.compareHashes(storedVerifierHash, verifierHash)
-      // if (!match) {
-      //   // if hash does not match, 401 and delete token
-      //   await AccountUtilities.deleteSelector(resetToken)
-      //   ctx.assert(match, 401)
-      // }
+      await Methods.sendMagicLinkEmail(email, emailToken)
+      ctx.body = { ok: true }
     } catch (e) {
-      ctx.throw(401)
+      console.log(e)
+      ctx.body = { ok: false, message: e.message || 'unknown error' }
     }
+  }
+
+  // Endpoint to create hash to send in email to the user on register
+  public static async register (ctx: BaseContext) {
+    const { email, role } = ctx.request.body
+    ctx.assert(email, 400, Strings.EmailIsRequired)
+    ctx.assert(role, 400, Strings.RoleIsRequired)
 
-    // if match and hasn't expired, reset the password
-    return AccountController.resetPassword(ctx)
+    // Generate token, store it
+    const emailToken = await generateAndStoreToken(ctx, false)
+    try {
+      await Methods.sendMagicLinkEmail(email, emailToken)
+      ctx.body = { ok: true }
+    } catch (e) {
+      ctx.body = { ok: false, message: e.message || 'unkonwn error' }
+    }
   }
-  public static async resetPassword (ctx: BaseContext) {
-    ctx.status = 200
-    ctx.body = { ok: true }
+
+  // Endpoint to verify hash from email magic link and send back auth token
+  public static async verifyAuth (ctx: BaseContext) {
+    const { email, auth } = ctx.request.body
+    ctx.assert(email, 400, Strings.EmailIsRequired)
+    ctx.assert(auth, 400, Strings.AuthIsRequired)
+
+    // Look up hash sent in request in the db 
+    try {
+      const token = await getStoredVerifierHash(ctx.request.body.auth)
```

Comment:
I think this call returns the row from Dynamo (hence `token.email` on next line) so don't call it `token`. 

What does the body of this request look like? `getStoredVerifierHash` is expecting an `AuthToken` as input, which is an object of form `{ selector: Buffer, verifier: Buffer }`. I think some more parsing is necessary here. I expect the post body to look like this: `{ email, auth }` where `auth` is a Base64 string of the selector and verifier buffers concat'd (looks like that's what gets sent to them in the email). In that case, you can use the `splitSelectorVerifier` function to get `{ selector, verifier }` out of it. Then you'll get the row back from Dynamo based on the selector.

When we get this row back, we should be checking to see if the verifier hash matches. You'll have to see if the DocumentClient automatically converts the returned verifier hash to a buffer. If not, it will be `Buffer.from(token.verifierHash, 'base64')`. Use the `compareHashes` function already written to compare the returned verifierHash buffer with the verifier they sent up in their request.

Reply:
Realized that I missed a step in this comment. The client is sending up `Buffer.concat([selector, verifier]).toString('base64')`. After we call `splitSelectorVerifier` we get `{ selector, verifier }`, which can then be passed to `getVerifierHash`. The return val from that function is what should be compared to the returned verifierHash from Dynamo.

Summary:
- user requests link
- we generate s, v
- we store s, hash(v) in dynamo
- we send s, v to user as sv
- when they attempt to verify, they send up s2v2
- we split s2v2 into s2, v2
- we look up s2 in dynamo and get back hash(v)
- we compute hash(v2) and compare it to hash(v)

Reply:
i _think_ this is done